integration-tests/src/test/java/tachyon/master/JournalCrashIntegrationTest.java

```diff
@@ -0,0 +1,232 @@
+import tachyon.client.TachyonFS;
+import tachyon.conf.TachyonConf;
+import tachyon.thrift.FileDoesNotExistException;
+import tachyon.thrift.InvalidPathException;
+import tachyon.thrift.TableDoesNotExistException;
+import tachyon.util.CommonUtils;
+
+/**
+ * Test master journal for crashes. Assert that test can replay the editlog and reproduce the
+ * correct state. Test both the single master(tachyon) and multi masters(tachyon-ft) mode.
+ */
+public class JournalCrashIntegrationTest {
+
+  /**
+   * Hold a client and keep creating files/tables.
+   */
+  class ClientThread implements Runnable {
+    /** The number of successfully created files/tables. */
+    public int mSuccessNum = 0;
+
+    private final int mOpType;  //0:create file; 1:create raw table.
+    private final TachyonFS mTfs;
+
+    public ClientThread(int opType, TachyonFS tfs) {
+      mOpType = opType;
+      mTfs = tfs;
+    }
+
+    /**
+     * Keep creating files/tables until something crashes or fail to create. Record how many files/
+     * tables are created successfully.
+     */
+    @Override
+    public void run() {
+      try {
+        // This infinity loop will be broken if something crashes or fail to create. This is
+        // expected since we are testing the crash scenario.
+        while (true) {
+          if (mOpType == 0) {
+            if (mTfs.createFile(new TachyonURI(TEST_FILE_DIR + mSuccessNum)) == -1) {
+              break;
+            }
+          } else if (mOpType == 1) {
+            if (mTfs.createRawTable(new TachyonURI(TEST_TABLE_DIR + mSuccessNum), 1) == -1) {
+              break;
+            }
+          }
+          mSuccessNum ++;
+          CommonUtils.sleepMs(null, 100);
+        }
+      } catch (Exception e) {
+        // Something crashed. Stop the thread.
+      }
+    }
+  }
+
+  private static final int TEST_BLOCK_SIZE = 128;
+  private static final String TEST_FILE_DIR = "/files/";
+  private static final int TEST_MASTERS = 3;
+  private static final String TEST_TABLE_DIR = "/tables/";
+  private static final long TEST_TIME_MS = Constants.SECOND_MS;
+
+  private ClientThread mCreateFileThread = null;
+  private ClientThread mCreateTableThread = null;
+  /** Executor for running client threads */
+  private final ExecutorService mExecutorsForClient = Executors.newFixedThreadPool(2);
+  /** Executor for constructing MasterInfo */
+  private final ExecutorService mExecutorsForMasterInfo = Executors.newFixedThreadPool(2);
+  /** Put all the threads of local cluster in this group then they can be killed at any time */
+  private ThreadGroup mLocalClusterThreadGroup = null;
+  private TachyonConf mMasterTachyonConf =  null;
+
+  @After
+  public final void after() throws Exception {
+    mExecutorsForClient.shutdown();
+    mExecutorsForMasterInfo.shutdown();
+    mLocalClusterThreadGroup.interrupt();
+    // Clear system properties since the cluster is not stopped gracefully.
+    System.clearProperty(Constants.TACHYON_HOME);
```

Comment:
I think line 108-121 are from some legacy code.  It is fine to have them here for this PR. I can clean it up later.